adapters/repos/db/shard_write_batch_objects.go

```diff
@@ -0,0 +1,273 @@
+	batchId int, objectIndex int, object *storobj.Object) error {
+	if _, ok := b.duplicates[batchId+objectIndex]; ok {
+		return nil
+	}
+	uuidParsed, err := uuid.Parse(object.ID().String())
+	if err != nil {
+		return errors.Wrap(err, "invalid id")
+	}
+
+	idBytes, err := uuidParsed.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	status, err := b.shard.putObjectInTx(tx, object, idBytes)
+	if err != nil {
+		return err
+	}
+
+	b.setStatusForID(status, object.ID())
+
+	if err := ctx.Err(); err != nil {
+		return errors.Wrapf(err, "end transaction %d of batch", batchId)
+	}
+	return nil
+}
+
+// setStatusForID is thread-safe as it uses the underlying mutex to lock the
+// statuses map when writing into it
+func (b *objectsBatcher) setStatusForID(status objectInsertStatus, id strfmt.UUID) {
+	b.Lock()
+	b.statuses[id] = status
+	b.Unlock()
+}
+
+// storeObjectsAdditionalStorage stores the object in all non-key-value stores,
+// such as the main vector index as well as the property-specific indices, such
+// as the geo-index.
+func (b *objectsBatcher) storeObjectsAdditionalStorage(ctx context.Context) {
+	if ok := b.checkContext(ctx); !ok {
+		// if the context is no longer OK, there's no point in continuing - abort
+		// early
+		return
+	}
+
+	beforeVectorIndex := time.Now()
+	wg := &sync.WaitGroup{}
+	for i, object := range b.objects {
+		if b.shouldSkipInAdditionalStorage(i) {
+			continue
+		}
+
+		wg.Add(1)
+		status := b.statuses[object.ID()]
+		go func(object *storobj.Object, status objectInsertStatus, index int) {
+			defer wg.Done()
+			b.storeSingleObjectInAdditionalStorage(ctx, object, status, index)
+		}(object, status, i)
+	}
+	wg.Wait()
+	b.shard.metrics.VectorIndex(beforeVectorIndex)
+}
+
+func (b *objectsBatcher) shouldSkipInAdditionalStorage(i int) bool {
+	b.Lock()
+	_, ok := b.errs[i]
+	b.Unlock()
+	if ok {
+		// had an error prior, ignore
+		return true
+	}
+
+	// no need to lock the mutex for a duplicate check, as we only ever write
+	// during init() in there - not concurrently
+	if _, ok := b.duplicates[i]; ok {
+		// is a duplicate, ignore
+		return true
+	}
+
+	return false
+}
+
+func (b *objectsBatcher) storeSingleObjectInAdditionalStorage(ctx context.Context,
+	object *storobj.Object, status objectInsertStatus, index int) {
+	if err := ctx.Err(); err != nil {
+		b.setErrorAtIndex(errors.Wrap(err, "insert to vector index"), index)
+		return
+	}
+
+	if err := b.shard.updateVectorIndex(object.Vector, status); err != nil {
+		b.setErrorAtIndex(errors.Wrap(err, "insert to vector index"), index)
+		return
+	}
+
+	if err := b.shard.updatePropertySpecificIndices(object, status); err != nil {
+		b.setErrorAtIndex(errors.Wrap(err, "update prop-specific indices"), index)
+		return
+	}
+}
+
+// setErrorAtIndex is thread-safe as it uses the underlying mutex to lock
+// writing into the errs map
+func (b *objectsBatcher) setErrorAtIndex(err error, index int) {
+	b.Lock()
+	b.errs[index] = err
+	b.Unlock()
```

Comment:
I would also here change that line to `defer b.Unlock()` and move it just after the `b.Lock()` so that we would always unlock